File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Maximum_Twin_Sum_of_a_Linked_List/yujin.cpp
/**
 * Definition for singly-linked list.
 */
#include <vector>
#include <stdlib.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    int pairSum(ListNode* head) {
        vector<int> nodeArray;
        ListNode* cur = head;
        while (cur != nullptr) {
            nodeArray.push_back(cur->val);
            cur = cur->next;
        }

        int maxTwinSum = 0;
        for (int i = 0; i < nodeArray.size() / 2; i++) {
            int twinSum = nodeArray[i] + nodeArray[nodeArray.size() / 2 - 1 - i];
            if (maxTwinSum < twinSum) { maxTwinSum = twinSum; }
        }
        return maxTwinSum;
    }
};

File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Maximum_Twin_Sum_of_a_Linked_List/yujin.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        node_val_list = []
        cur = head
        while cur:
            node_val_list.append(cur.val)
            cur = cur.next
        
        max_twin_sum = 0
        for i in range(len(node_val_list) // 2):
            twin_sum = node_val_list[i] + node_val_list[len(node_val_list) - 1 - i]
            if max_twin_sum < twin_sum:
                max_twin_sum = twin_sum

        return max_twin_sum
            
